server/PendingOps.hs

```diff
@@ -0,0 +1,174 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- | This module implements a secure and efficient way to temporarily store
+-- webauthn options for the create/register and get/login webauthn operations
+-- along with generating the challenge needed for them.
+module PendingOps
+  ( newPendingOps,
+    PendingOps,
+    insertPendingOptions,
+    getPendingOptions,
+  )
+where
+
+import Control.Concurrent (forkIO, threadDelay)
+import qualified Control.Concurrent.STM as STM
+import Control.Monad (forever, unless)
+import qualified Crypto.Fido2.Model as M
+import Crypto.Fido2.Model.WebauthnType (SWebauthnType (SCreate, SGet), SingI, sing)
+import Data.Binary (Binary (get, put))
+import qualified Data.Binary as Binary
+import Data.Binary.Get as Binary (getInt64le, getRemainingLazyByteString)
+import Data.Binary.Put as Binary (putInt64le, putLazyByteString)
+import qualified Data.ByteString.Lazy as LBS
+import Data.Int (Int64)
+import Data.Map (Map)
+import qualified Data.Map as Map
+import System.Clock (Clock (Monotonic), TimeSpec (sec), getTime)
+import System.Random.Stateful (globalStdGen, uniformByteStringM)
+
+-- | An 'M.Challenge' that expires after a certain time.
+data ExpiringChallenge = ExpiringChallenge
+  { -- | The system time in seconds before which this challenge certainly isn't expired
+    -- This value originates from @'getTime' Monotonic@, so its absolute value
+    -- is only meaningful to the system that generated it
+    -- This field intentionally comes _before_ the 'randomness' field, such that
+    -- the derived 'Ord' instance orders challenges according to their expiration
+    -- time, which allows much faster periodic expiration
+    notExpiredBefore :: Int64,
```

Comment:
I now switched to wall time for this, because https://security.stackexchange.com/questions/187316/is-exposing-the-server-time-a-security-risk. Added comments mentioning this